src/iuse.cpp

```diff
@@ -2117,30 +2163,212 @@ int iuse::sew_advanced(player *p, item *it, bool, const tripoint& )
-        if( mod->inc_damage() ) {
-            p->add_msg_if_player(m_bad, _("You destroy it!"));
-            p->i_rem_keep_contents( pos );
-        }
-        return thread_needed / 2;
+
+		filament_left = filament_needed / 2;
+
+		// Go through all filament items and spend available until finished
+		for (auto &filament : mod_filament) {
+
+			auto inv = p->crafting_inventory();
+			auto itm = inv.item_by_type(filament);
+
+			bool filament_count_by_charges = itm.count_by_charges();
+
+			filament_have = (filament_count_by_charges) ? crafting_inv.charges_of(filament) : crafting_inv.amount_of(filament);
+
+			if (filament_have == filament_left)
+			{
+				filament_used = filament_have;
+			}
+			else if (filament_have > filament_left)
+			{
+				filament_used = filament_left;
+			}
+			else if (filament_have < filament_left)
+			{
+				filament_used = filament_have;
+			};
+
+			if (filament_count_by_charges)
+			{
+				p->use_charges(filament, filament_used);
+			}
+			else
+			{
+				p->use_amount(filament, filament_used);
+			};
+
+			filament_left -= filament_used;
+
+			//p->add_msg_if_player(m_info, _("DEBUG: Used filament=%s, quantity=%d, quantity_used=%d, quantity_left=%d"), filament, filament_have, filament_used, filament_left);
+
+			if (filament_left == 0)
+			{
+				break;
+			};
+		};
+
+		p->add_msg_if_player(m_bad, _("You damage your %s trying to modify it and also waste filament (%d)!"),
+			mod->tname().c_str(), filament_needed / 2);
+		if (mod->inc_damage()) {
+			p->add_msg_if_player(m_bad, _("You destroy it!"));
+			p->i_rem_keep_contents(pos);
+		}
+
+		return 0;
+
     } else if( rn <= 10 ) {
-        p->add_msg_if_player( m_bad,
-                              _("You fail to modify the clothing, and you waste thread and materials.") );
-        p->consume_items( comps );
-        return thread_needed;
+
+		filament_left = filament_needed;
+
+		// Go through all filament items and spend available until finished
+		for (auto &filament : mod_filament) {
+
+			auto inv = p->crafting_inventory();
+			auto itm = inv.item_by_type(filament);
+
+			bool filament_count_by_charges = itm.count_by_charges();
+
+			filament_have = (filament_count_by_charges) ? crafting_inv.charges_of(filament) : crafting_inv.amount_of(filament);
+
+			if (filament_have == filament_left)
+			{
+				filament_used = filament_have;
+			}
+			else if (filament_have > filament_left)
+			{
+				filament_used = filament_left;
+			}
+			else if (filament_have < filament_left)
+			{
+				filament_used = filament_have;
+			};
+
+			if (filament_count_by_charges)
+			{
+				p->use_charges(filament, filament_used);
+			}
+			else
+			{
+				p->use_amount(filament, filament_used);
+			};
+
+			filament_left -= filament_used;
+
+			//p->add_msg_if_player(m_info, _("DEBUG: Used filament=%s, quantity=%d, quantity_used=%d, quantity_left=%d"), filament, filament_have, filament_used, filament_left);
+
+			if (filament_left == 0)
+			{
+				break;
+			};
+		};
```

Comment:
Repeating a whole block like that is not acceptable.
Just do not return at the end of the block and instead have the consumption happen after all the `if/else` here.